src/dnm/c14n.rs

```diff
@@ -0,0 +1,120 @@
+//! The `dnm::c14n` submodule offers lightweight canonicalization for a DOM node.
+//!      The core purpose for canonicalization is linguistic comparison,
+//!      so the c14n module tries to strip away markup artefacts unrelated to the underlying content, such as xml:ids.
+use libxml::tree::Node;
+use libxml::tree::NodeType::{TextNode, ElementNode};
+use crypto::md5::Md5;
+use crypto::digest::Digest;
+use std::str;
+use std::sync::Mutex;
+use dnm::DNM;
+
+lazy_static! {
+  static ref MD5_HASHER : Mutex<Md5> = Mutex::new(Md5::new());
+}
+
+impl DNM {
+  /// Our linguistic canonical form will only include 1) node name, 2) class attribute and 3) textual content
+  ///  - excludes certain experimental markup, such as all math annotation elements
+  ///  - excludes whitespace nodes and comment nodes
+  pub fn to_c14n_basic(&self) -> String {
+    self.node_c14n_basic(&self.root_node)
+  }
+
+  /// Canonicalize a single node of choice
+  pub fn node_c14n_basic(&self, node: &Node) -> String {
+    let mut canonical_node = String::new();
+    self.canonical_internal(node, None, &mut canonical_node);
+    canonical_node
+  }
+
+  /// Obtain an MD5 hash from the canonical string of the entire DOM
+  pub fn to_hash_basic(&self) -> String {
+    self.node_hash_basic(&self.root_node)
+  }
+
+  /// Obtain an MD5 hash from the canonical string of a Node
+  pub fn node_hash_basic(&self, node: &Node) -> String {
+    let mut hasher = MD5_HASHER.lock().unwrap();
+    hasher.reset();
+    hasher.input_str(&self.node_c14n_basic(node));
+    hasher.result_str()
+  }
+
+  fn canonical_internal(&self, node: &Node, indent : Option<u32>, mut canonical_node : &mut String) {
+    // Bookkeep indents, if requested
+    let indent_string = match indent {
+      Some(level) => String::new() + "\n" + &(1..level).map(|_| " ").collect::<String>(),
+      None => String::new()
+    };
+    let next_indent_level = match indent {
+      Some(level) => Some(level + 2),
+      None => None
+    };
+
+    match node.get_type() {
+      Some(TextNode) => {
+        if let Ok(range) = self.get_range_of_node(node) {
```

Comment:
The main reason I tied the c14n to the DNM is here. I wanted to make use of the plain text parameters we already support without extra effort, and by hooking the plain text lookup to the DNM, it is very easy to ask for unidecode, stemming, lower-casing etc.